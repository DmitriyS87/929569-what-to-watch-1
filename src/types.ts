export enum Tab {
  OWERVIEW = 'Overview',
  DETAILS = 'Details',
  REVIEWS = 'Reviews',
}

export interface Movie {
  title: string;
  poster: {
    src: string;
    alt: string;
  };
  previewImg: {
    src: string;
    alt: string;
  };
  id: number;
  previewMovie: {
    href: string;
    format: string;
  }[];
  genre: string;
  director: string;
  description: string;
  releseYear: number;
  runTime: number;
  rating: number;
  isFavorite: boolean;
  scoresCount: number;
  starring: string[];
  fullMovie: {
    href: string;
    format: string;
  };
  backgroundColor: string;
  backgroundImg: { src: string; alt: string };
}

export interface Review {
  id: number;
  user: {
    id: number;
    name: string;
  };
  rating: number;
  comment: string;
  date: string;
}

interface RootState {
  movies: Movie[];
}

interface RootState {
  isAuthorizationRequired: boolean;
  user: null | {};
  errorMessage: string;
}

interface RootState {
  genre: string;
}

export { RootState };
