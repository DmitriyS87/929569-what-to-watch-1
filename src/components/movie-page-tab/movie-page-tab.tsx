import * as React from 'react';
import { Movie, Review } from '../../types';
import { TABS } from '../../constants/movie-page-tab.constant';
import * as moment from 'moment';
import { connect } from 'react-redux';
import { Operation as OperationData } from '../../reducers/data/data.js';
import { getStringRating } from '../../utils/get-string-rating';
import { getMovieLengthFormat } from '../../utils/get-movie-length-format';

interface Props {
  activeItem: string;
  movie: Movie;
  comments: Review[] | null;
  isLoading: boolean;
  loadMovieComments: (id: number) => void;
}

interface PropsNav {
  setActive: (param: string) => void;
}

const MoviePageTab = (props: Props & PropsNav) => {
  const { activeItem, setActive, movie, comments, isLoading, loadMovieComments } = props;
  const handleClick = evt => {
    evt.preventDefault();
    setActive(evt.target.textContent);
  };

  const ScreenTab = (props: Props) => {
    const {
      genre,
      releseYear,
      director,
      starring,
      runTime,
      rating,
      scoresCount,
      description,
      id,
    } = props.movie;
    const { comments, isLoading, loadMovieComments } = props;
    const descriptions = description.split('. ');

    if (!isLoading && !comments) {
      loadMovieComments(id);
    }

    const sortComentsByDate = (reviews: Review[]): Review[] => {
      return reviews.sort((a, b) => {
        if (Date.parse(a.date) > Date.parse(b.date)) {
          return -1;
        }
        if (Date.parse(a.date) < Date.parse(b.date)) {
          return 1;
        }
        return 0;
      });
    };

    const halveComments = (comments: Review[]): { left: Review[]; right: Review[] } => {
      const columns: { left: Review[]; right: Review[] } = {
        left: [],
        right: [],
      };
      let left = false;
      const toggleLeft: () => void = () => {
        left = !left;
      };
      for (let comment of comments) {
        toggleLeft();
        if (left) {
          columns.left.push(comment);
        }
        if (!left) {
          columns.right.push(comment);
        }
      }
      return columns;
    };

    const computedComments: any = isLoading
      ? { left: [`Loading... Please wait`], right: [] }
      : comments === null
      ? { left: [`no coments here. Would you like to smthn to type?! Be first!`], right: [] }
      : halveComments(sortComentsByDate(comments));

    const makeReview = (reviewData, key) => {
      const reviewDate = Date.parse(reviewData.date);
      return (
        <div key={key} className='review'>
          <blockquote className='review__quote'>
            <p className='review__text'>{reviewData.comment}</p>

            <footer className='review__details'>
              <cite className='review__author'>{reviewData.user.name}</cite>
              <time className='review__date' dateTime={moment(reviewDate).format(`YYYY-MM-DD`)}>
                {moment(reviewDate).format(`MMMM D, YYYY`)}
              </time>
            </footer>
          </blockquote>

          <div className='review__rating'>{reviewData.rating}</div>
        </div>
      );
    };

    switch (props.activeItem) {
      case `Overview`:
        return (
          <React.Fragment>
            <div className='movie-rating'>
              <div className='movie-rating__score'>{rating.toFixed(1)}</div>
              <p className='movie-rating__meta'>
                <span className='movie-rating__level'>{getStringRating(rating)}</span>
                <span className='movie-rating__count'>{scoresCount} ratings</span>
              </p>
            </div>

            <div className='movie-card__text'>
              {descriptions.map((paragraph, idx) => {
                return <p key={`DescriptionMovie_${idx}`}>{paragraph}</p>;
              })}
              <p className='movie-card__director'>
                <strong>Director: {director}</strong>
              </p>
              <p className='movie-card__starring'>
                <strong>
                  Starring:{' '}
                  {starring.map((it, idx, arr) => `${it}${idx < arr.length - 1 ? ',' : ''}`)} and
                  other
                </strong>
              </p>
            </div>
          </React.Fragment>
        );

      case `Details`:
        return (
          <div className='movie-card__text movie-card__row'>
            <div className='movie-card__text-col'>
              <p className='movie-card__details-item'>
                <strong className='movie-card__details-name'>Director</strong>
                <span className='movie-card__details-value'>{director}</span>
              </p>
              <p className='movie-card__details-item'>
                <strong className='movie-card__details-name'>Starring</strong>
                <span className='movie-card__details-value'>
                  {starring.map((artist, idx, array) => {
                    if (idx === array.length - 1) {
                      return `${artist}`;
                    }
                    return (
                      <React.Fragment key={`artist${artist}${idx}`}>
                        {`${artist}`} <br />
                      </React.Fragment>
                    );
                  })}
                </span>
              </p>
            </div>

            <div className='movie-card__text-col'>
              <p className='movie-card__details-item'>
                <strong className='movie-card__details-name'>Run Time</strong>
                <span className='movie-card__details-value'>{getMovieLengthFormat(runTime)}</span>
              </p>
              <p className='movie-card__details-item'>
                <strong className='movie-card__details-name'>Genre</strong>
                <span className='movie-card__details-value'>{genre}</span>
              </p>
              <p className='movie-card__details-item'>
                <strong className='movie-card__details-name'>Released</strong>
                <span className='movie-card__details-value'>{releseYear}</span>
              </p>
            </div>
          </div>
        );
      case `Reviews`:
        return (
          <div className='movie-card__reviews movie-card__row'>
            <div className='movie-card__reviews-col'>
              {computedComments.left.map((comment, idx) =>
                makeReview(comment, `computedCommentsleft${idx}`)
              )}
            </div>
            <div className='movie-card__reviews-col'>
              {computedComments.right.map((comment, idx) =>
                makeReview(comment, `computedCommentsRight${idx}`)
              )}
            </div>
          </div>
        );
      default:
        return <div />;
    }
  };

  return (
    <React.Fragment>
      <div className='movie-card__desc'>
        <nav className='movie-nav movie-card__nav'>
          <ul className='movie-nav__list'>
            {TABS.map((it, idx) => {
              return (
                <li
                  key={`MNL-${it}-${idx}`}
                  className={`movie-nav__item ${
                    it === activeItem ? `movie-nav__item--active` : ``
                  }`}
                >
                  <a href='#' onClick={handleClick} className='movie-nav__link'>
                    {it}
                  </a>
                </li>
              );
            })}
          </ul>
        </nav>

        <ScreenTab
          movie={movie}
          activeItem={activeItem}
          comments={comments}
          isLoading={isLoading}
          loadMovieComments={loadMovieComments}
        />
      </div>
    </React.Fragment>
  );
};

const mapDispatchToProps = dispatch => {
  return {
    loadMovieComments: id => dispatch(OperationData.getMovieComments(id)),
  };
};

export default connect(
  null,
  mapDispatchToProps
)(MoviePageTab);
export { MoviePageTab };
