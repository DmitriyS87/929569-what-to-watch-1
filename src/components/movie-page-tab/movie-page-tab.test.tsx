import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { MoviePageTab } from './movie-page-tab';
import mockData from '../../mocks/test-mocks.js';
import { Tab } from '../../types';

const mockReview = {
  id: 0,
  user: {
    id: 10,
    name: `Bob Marley`,
  },
  rating: 4.7,
  comment: `The best movie i have seen ever! All people should see it! Awsome! Bravo! Just see it now! It was a best lifetime!`,
  date: `2019-05-08T14:13:56.569Z`,
};

it(`MoviePageTab correctly renders Tab.OWERVIEW after relunch: `, () => {
  const clickHandler = jest.fn();
  const tree = renderer
    .create(
      <MoviePageTab
        activeItem={Tab.OWERVIEW}
        setActive={clickHandler}
        movie={mockData.film}
        comments={[mockReview]}
        isLoading={false}
        loadMovieComments={clickHandler}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it(`MoviePageTab correctly renders Tab.DETAILS after relunch: `, () => {
  const clickHandler = jest.fn();
  const tree = renderer
    .create(
      <MoviePageTab
        activeItem={Tab.DETAILS}
        setActive={clickHandler}
        movie={mockData.film}
        comments={[mockReview]}
        isLoading={false}
        loadMovieComments={clickHandler}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it(`MoviePageTab correctly renders Tab.REVIEWS after relunch: `, () => {
  const clickHandler = jest.fn();
  const tree = renderer
    .create(
      <MoviePageTab
        activeItem={Tab.REVIEWS}
        setActive={clickHandler}
        movie={mockData.film}
        comments={[mockReview]}
        isLoading={false}
        loadMovieComments={clickHandler}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
