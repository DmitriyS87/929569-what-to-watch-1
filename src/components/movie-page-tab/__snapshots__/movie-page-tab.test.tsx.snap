// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviePageTab correctly renders Tab.DETAILS after relunch:  1`] = `
<div
  className="movie-card__desc"
>
  <nav
    className="movie-nav movie-card__nav"
  >
    <ul
      className="movie-nav__list"
    >
      <li
        className="movie-nav__item "
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Overview
        </a>
      </li>
      <li
        className="movie-nav__item movie-nav__item--active"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Details
        </a>
      </li>
      <li
        className="movie-nav__item "
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Reviews
        </a>
      </li>
    </ul>
  </nav>
  <div
    className="movie-card__text movie-card__row"
  >
    <div
      className="movie-card__text-col"
    >
      <p
        className="movie-card__details-item"
      >
        <strong
          className="movie-card__details-name"
        >
          Director
        </strong>
        <span
          className="movie-card__details-value"
        >
          Test person
        </span>
      </p>
      <p
        className="movie-card__details-item"
      >
        <strong
          className="movie-card__details-name"
        >
          Starring
        </strong>
        <span
          className="movie-card__details-value"
        >
          Best Artist
           
          <br />
          Best Acter
           
          <br />
          Best Person
        </span>
      </p>
    </div>
    <div
      className="movie-card__text-col"
    >
      <p
        className="movie-card__details-item"
      >
        <strong
          className="movie-card__details-name"
        >
          Run Time
        </strong>
        <span
          className="movie-card__details-value"
        >
          4h 25m
        </span>
      </p>
      <p
        className="movie-card__details-item"
      >
        <strong
          className="movie-card__details-name"
        >
          Genre
        </strong>
        <span
          className="movie-card__details-value"
        >
          Adventure
        </span>
      </p>
      <p
        className="movie-card__details-item"
      >
        <strong
          className="movie-card__details-name"
        >
          Released
        </strong>
        <span
          className="movie-card__details-value"
        >
          1917
        </span>
      </p>
    </div>
  </div>
</div>
`;

exports[`MoviePageTab correctly renders Tab.OWERVIEW after relunch:  1`] = `
<div
  className="movie-card__desc"
>
  <nav
    className="movie-nav movie-card__nav"
  >
    <ul
      className="movie-nav__list"
    >
      <li
        className="movie-nav__item movie-nav__item--active"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Overview
        </a>
      </li>
      <li
        className="movie-nav__item "
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Details
        </a>
      </li>
      <li
        className="movie-nav__item "
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Reviews
        </a>
      </li>
    </ul>
  </nav>
  <div
    className="movie-rating"
  >
    <div
      className="movie-rating__score"
    >
      15.0
    </div>
    <p
      className="movie-rating__meta"
    >
      <span
        className="movie-rating__level"
      >
        
      </span>
      <span
        className="movie-rating__count"
      >
        542
         ratings
      </span>
    </p>
  </div>
  <div
    className="movie-card__text"
  >
    <p>
      Test text about current movie
    </p>
    <p>
      Next row test text about current movie
    </p>
    <p>
      Last row test text about current movie.
    </p>
    <p
      className="movie-card__director"
    >
      <strong>
        Director: 
        Test person
      </strong>
    </p>
    <p
      className="movie-card__starring"
    >
      <strong>
        Starring:
         
        Best Artist,
        Best Acter,
        Best Person
         and other
      </strong>
    </p>
  </div>
</div>
`;

exports[`MoviePageTab correctly renders Tab.REVIEWS after relunch:  1`] = `
<div
  className="movie-card__desc"
>
  <nav
    className="movie-nav movie-card__nav"
  >
    <ul
      className="movie-nav__list"
    >
      <li
        className="movie-nav__item "
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Overview
        </a>
      </li>
      <li
        className="movie-nav__item "
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Details
        </a>
      </li>
      <li
        className="movie-nav__item movie-nav__item--active"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Reviews
        </a>
      </li>
    </ul>
  </nav>
  <div
    className="movie-card__reviews movie-card__row"
  >
    <div
      className="movie-card__reviews-col"
    >
      <div
        className="review"
      >
        <blockquote
          className="review__quote"
        >
          <p
            className="review__text"
          >
            The best movie i have seen ever! All people should see it! Awsome! Bravo! Just see it now! It was a best lifetime!
          </p>
          <footer
            className="review__details"
          >
            <cite
              className="review__author"
            >
              Bob Marley
            </cite>
            <time
              className="review__date"
              dateTime="2019-05-08"
            >
              May 8, 2019
            </time>
          </footer>
        </blockquote>
        <div
          className="review__rating"
        >
          4.7
        </div>
      </div>
    </div>
    <div
      className="movie-card__reviews-col"
    />
  </div>
</div>
`;
